### ISSUE TITLE
feat: count deprecated API usage in metrics

### ISSUE THREAD

**Issue by seven332** [seven332/bangbang#783]
## Background

Firecracker exposes a `deprecated_api.deprecated_http_api_calls` metric for supported request shapes that use deprecated API fields. bangbang already parses several compatible deprecated fields, but its metrics output does not expose or update the matching Firecracker metric.

Relevant Firecracker examples include:

- `PUT /machine-config` and `PATCH /machine-config` with `cpu_template` present.
- `PUT /mmds/config` using MMDS V1.
- `PUT /vsock` with the deprecated `vsock_id` field present.
- `PUT /snapshot/load` with deprecated `mem_file_path` or `enable_diff_snapshots`.

## Problem Or Requirement

Expose and update the Firecracker-compatible deprecated API counter for valid parsed requests that include deprecated fields already recognized by bangbang. Malformed requests rejected before VMM dispatch should remain uncounted, matching the current metrics model for parser failures.

## Acceptance Criteria

- Metrics serialization includes `deprecated_api.deprecated_http_api_calls` when deprecated API usage is recorded.
- Valid parsed requests with deprecated `cpu_template`, MMDS V1 config, deprecated `vsock_id`, and deprecated snapshot-load fields increment the counter.
- Malformed requests that fail before typed request dispatch do not increment the deprecated API counter.
- Tests cover runtime metrics serialization and API socket behavior, including no echo of deprecated field values or host paths in metrics output.
- Compatibility docs and the validation matrix mention the deprecated API metric subset.

## Out Of Scope

- Adding Firecracker deprecation warning response messages.
- Implementing real snapshot create/load support.
- Implementing CPU templates, MMDS behavior changes, or vsock behavior changes beyond metrics accounting.

**Comment by seven332** [comment]
## Research Phase

## Scope

Issue #783 tracks Firecracker-compatible `deprecated_api.deprecated_http_api_calls` metrics for deprecated request fields that bangbang already parses. It does not track deprecation warning response text, real snapshot support, CPU template support, MMDS behavior changes, or vsock behavior changes.

## Repository Guidance Read

- `AGENTS.md`: Rust workspace commands, strict linting, review expectations, and integration-test boundaries.
- `docs/review-guidelines.md`: metrics/API changes need behavior tests, parser-failure coverage, redaction checks, and docs updates.
- `README.md`, `docs/firecracker-compatibility.md`, and `docs/firecracker-validation-matrix.md`: current compatibility status and metrics scope.
- Relevant code in `crates/api/src/http.rs`, `crates/bangbang/src/api_server.rs`, `crates/bangbang/src/vmm.rs`, `crates/runtime/src/lib.rs`, and `crates/runtime/src/metrics.rs`.

## Firecracker Facts

Firecracker defines `DeprecatedApiMetrics` with `deprecated_http_api_calls` in `src/vmm/src/logger/metrics.rs`.

Firecracker increments this metric for deprecated fields in current request parsing paths:

- `PUT /machine-config` when `cpu_template` is present.
- `PATCH /machine-config` when `cpu_template` is present.
- `PUT /mmds/config` when the requested MMDS version is `V1`.
- `PUT /vsock` when `vsock_id` is present.
- `PUT /snapshot/load` when deprecated `mem_file_path` is present or `enable_diff_snapshots` is true.

Firecracker's balloon request parser does not expose API request metrics fields for balloon endpoints in the current metrics struct.

## bangbang Facts

- `MetricsState` currently serializes GET, PUT, PATCH request metrics and logger metrics, but no `deprecated_api` section.
- `MachineConfigRequest::cpu_template()` and `MachineConfigPatchRequest::cpu_template()` expose whether `cpu_template` was present after parsing.
- `MmdsConfigRequest::version()` exposes the parsed MMDS version. The API parser default is `MmdsVersion::V1`.
- `VsockConfigRequest::vsock_id()` exposes whether the deprecated `vsock_id` field was present after parsing.
- `ApiRequest::PutSnapshotLoad` currently carries no parsed load metadata, so the deprecated snapshot-load fields are consumed during parsing and unavailable to `api_server` dispatch.
- Existing counted API request behavior is centralized in typed VMM request wrappers. Parser failures that never reach VMM dispatch remain uncounted.
- Existing metrics tests assert exact serialized JSON strings, so adding a new metrics section or fields requires focused fixture updates.

## Constraints

- Deprecated field values and host paths must not be emitted in metrics output or errors introduced by this work.
- Malformed requests should remain uncounted when parsing fails before typed dispatch.
- The PR should stay metrics-only and avoid changing API status codes, response bodies, or behavior of the underlying unsupported actions.
- Snapshot-load deprecated-field accounting may require a small parsed request type because the current `PutSnapshotLoad` variant discards the needed flags.

## Verification Surface

- Runtime metrics serialization tests in `crates/runtime/src/metrics.rs` can cover the `deprecated_api` JSON section.
- API socket tests in `crates/bangbang/src/api_server.rs` can cover valid deprecated requests, malformed non-counting behavior, and redaction.
- Existing metrics-focused tests may need expected JSON updates only where the new section is intentionally present.
- Documentation updates are expected in `docs/firecracker-compatibility.md` and `docs/firecracker-validation-matrix.md`.

**Comment by seven332** [comment]
## Options Phase

## Option A: Count deprecated usage at API dispatch after parsing

Core idea: keep metrics owned by the process/VMM controller, add a small `deprecated_api` metrics recorder, and call it from `api_server` when a successfully parsed `ApiRequest` contains a deprecated field. Extend only snapshot load parsing enough to preserve deprecated-field flags for dispatch.

Advantages:

- Preserves bangbang's current metrics ownership model.
- Keeps malformed parser failures uncounted because only parsed `ApiRequest` values are inspected.
- Keeps the change small and explicit at the API/VMM boundary.
- Avoids global metrics state in the parser crate.
- Allows snapshot load accounting by carrying a tiny parsed metadata struct instead of full snapshot behavior.

Risks and trade-offs:

- Deprecated checks are spread across API dispatch unless wrapped in a small helper.
- Snapshot load needs an `ApiRequest` shape change and related parser tests because the current variant is data-less.
- Does not add Firecracker-style deprecation warning response messages; that remains out of scope.

## Option B: Push deprecated accounting into typed VMM request wrappers

Core idea: extend `PutApiRequest` and `PatchApiRequest` wrappers with deprecated metadata, then record deprecated metrics alongside request/failure metrics during VMM request handling.

Advantages:

- Keeps more accounting in the VMM request layer used by counted endpoints.
- Can colocate deprecated recording with current request/failure accounting for machine-config, MMDS, and vsock.

Risks and trade-offs:

- Not all affected endpoints use typed wrappers today; snapshot actions currently route directly through `handle_action`.
- Deprecated metadata is not method-specific request/failure accounting, so pushing it into wrappers blurs wrapper purpose.
- Requires more wrapper surface changes than the metric needs.

## Option C: Count deprecated usage in the parser crate

Core idea: mirror Firecracker more closely by incrementing deprecated metrics while parsing deprecated fields.

Advantages:

- Closest to Firecracker's implementation location.
- Parser has exact knowledge of deprecated fields, including snapshot load.

Risks and trade-offs:

- Conflicts with bangbang's current crate boundary: `bangbang-api` is parser-only and does not own process metrics state.
- Would require global metrics, callback plumbing, or parser return metadata beyond the current request enum.
- Makes parser failures and dispatch counting more likely to diverge from the existing bangbang metrics model.

## Option D: Add general parser metadata for deprecation messages and metrics

Core idea: introduce a generic parsed-request metadata object that can carry deprecation flags or messages for any API request, then record metrics and possibly warnings from that metadata.

Advantages:

- Creates a foundation for future Firecracker deprecation warning parity.
- Avoids bespoke checks per request in `api_server`.

Risks and trade-offs:

- Larger than the current metrics-only issue.
- Requires broader API parser and dispatch plumbing.
- More likely to create an abstraction before the repository has more metadata use cases.

## Comparison

Option A is the smallest approach that satisfies the issue while preserving current bangbang architecture. Option B overfits VMM wrappers and misses direct snapshot action routing. Option C mirrors Firecracker internally but breaks the parser/runtime separation. Option D may be useful later for warning-message parity, but it is too broad for this PR.

**Comment by seven332** [comment]
## Plan Phase

## Goal And Chosen Approach

Use Option A: keep metrics in the runtime/VMM controller, and record `deprecated_api.deprecated_http_api_calls` in the API dispatch layer after a request has parsed successfully. Carry only the minimal snapshot-load deprecated-field flags needed for metrics.

## Implementation Tasks

1. Add `DeprecatedApiMetrics` to `crates/runtime/src/metrics.rs` with a saturating counter, emptiness check, accessor, and JSON serialization under `deprecated_api.deprecated_http_api_calls`.
2. Add `MetricsState` and `VmmController` record methods for one deprecated API call.
3. Add a small parsed snapshot-load request type in `crates/api/src/http.rs` that exposes whether the load request used deprecated fields. Change `ApiRequest::PutSnapshotLoad` to carry that type.
4. Add an `api_server` helper that inspects successfully parsed `ApiRequest` values and records one deprecated API call when:
   - `PUT /machine-config` has `cpu_template().is_some()`.
   - `PATCH /machine-config` has `cpu_template().is_some()`.
   - `PUT /mmds/config` has `version() == V1`.
   - `PUT /vsock` has `vsock_id().is_some()`.
   - `PUT /snapshot/load` reports deprecated `mem_file_path` or `enable_diff_snapshots` usage.
5. Call that helper after parsing and before endpoint dispatch so valid parsed requests are counted, while parser failures remain uncounted.
6. Add runtime metrics serialization tests for the new `deprecated_api` section.
7. Add API socket metrics tests covering valid deprecated requests, malformed non-counting cases, and no echo of deprecated values or host paths in metrics output.
8. Update compatibility documentation and the validation matrix to mention the implemented deprecated API metric subset.

## Files Likely To Change

- `crates/runtime/src/metrics.rs`
- `crates/runtime/src/lib.rs`
- `crates/api/src/http.rs`
- `crates/bangbang/src/api_server.rs`
- `docs/firecracker-compatibility.md`
- `docs/firecracker-validation-matrix.md`

## Compatibility Notes

- Count one deprecated API call per valid parsed request, not one per deprecated field.
- Keep deprecation warning response messages out of scope.
- Keep malformed parser failures uncounted.
- Do not change underlying machine-config, MMDS, vsock, or snapshot behavior.

## Validation

Run focused tests first, then the repository validation set required by `AGENTS.md` for API/runtime changes:

- `cargo fmt --all -- --check`
- `cargo check --workspace --all-targets --all-features --locked`
- `cargo test -p bangbang-runtime writes_deprecated_api_metrics_when_recorded --all-targets --all-features --locked`
- `cargo test -p bangbang configured_metrics_counts_deprecated_api_requests --all-targets --all-features --locked`
- `cargo test --workspace --all-targets --all-features --locked --exclude bangbang-hvf`
- `cargo test -p bangbang-hvf --lib --all-features --locked`
- `cargo clippy --workspace --all-targets --all-features --locked -- -D warnings`
- signed-target clippy commands listed in `AGENTS.md`
- `RUSTDOCFLAGS="-D warnings" cargo doc --workspace --all-features --no-deps --locked`
- `git diff --check`
- `scripts/run-signed-process-tests.sh`

## Definition Of Done

- Metrics output includes `deprecated_api.deprecated_http_api_calls` when deprecated API usage is recorded.
- Valid parsed deprecated-field requests increment the counter once per request.
- Malformed requests remain uncounted.
- Tests and docs cover the behavior and scope exclusions.
